### PR TITLE
Extend the AppView to be able to use helpers defined there.

### PR DESCRIPTION
**Pull request by dereuromark**
Resolves https://github.com/FriendsOfCake/CakePdf/issues/145

In AppView for example:

```
public function initialize() {
    $this->loadHelper('Tools.AuthUser');
    ...
}
```

Those helpers are also expected in the ctps in /pdf subfolder as otherwise it throws missing helper exceptions.

### CODE REVIEW

**Comment by dereuromark** [comment]
Any idea how I can get the AppView to be recognized for the tests?
const APP points to test_app already...

**Comment by ADmad** [comment]
Put the AppView.php under `tests/test_app/TestApp/View/`

**Comment by ADmad** [comment]
Hmm looking at the constants in `tests/bootstrap.php` it should be `tests/test_app/View/AppView.php`

File: src/View/PdfView.php [hunk 1]
```diff
@@ -12,18 +12,18 @@
...
 use Cake\Network\Request;
 use Cake\Network\Response;
-use Cake\View\View;
 
 /**
- * @package       Cake.View
+ * Renders PDFs instead of HTML.
  */
-class PdfView extends View
+class PdfView extends AppView
```
Comment (ADmad) [inline src/View/PdfView.php]: AppView is optional. This change can break existing apps which don't use AppView.

Comment (dereuromark) [inline src/View/PdfView.php]: Yes Requires major bump.

Comment (ADmad) [inline src/View/PdfView.php]: While it's added in official skeleton fact remains it's optional. It should have probably made it required in 3.2 but that boat has sailed.

Comment (dereuromark) [inline src/View/PdfView.php]: Yes, we will have to wait for 3.3: https://github.com/cakephp/cakephp/issues/8155
But we can already require it for this plugin in a new version, as long as we adjust the documentation as well.

**Comment by dereuromark** [comment]
As for the test fail:
Exactly, it is under `tests/test_app/View/AppView.php` where it should be.
I also tried `tests/test_app/src/View/AppView.php` as this is where I usually put my classes.
But both are not found :/

File: src/View/PdfView.php (hunk 1 above)
Comment (ADmad) [inline src/View/PdfView.php]: Yup with a version bump we can make that required.

Comment (dereuromark) [inline src/View/PdfView.php]: I added docs: https://github.com/FriendsOfCake/CakePdf/commit/2819ca7e26151c02ed68e6f905fadeedccce563b

**Comment by ADmad** [comment]
We could also take the route taken in FriendsOfCake/bootstrap-ui#110 and put all the important bits in a trait.

**Comment by dereuromark** [comment]
Maybe that is a better idea.
If anyone has time please adjust the PR or make a new one.

**Comment by ADmad** [comment]
@dereuromark Waiting up on others won't do any good :)  If you need the feature you will have to put in the legwork yourself.

**Comment by dereuromark** [comment]
So what did we decide to do now?
We can easily extend AppView in a future version (major bump?) - I already do it with https://github.com/dereuromark/cakephp-cache/blob/master/src/View/PartialCacheView.php#L19 successfully.
Testing is also no big deal using class_alias(): https://github.com/dereuromark/cakephp-cache/blob/master/tests/bootstrap.php#L76

**Comment by dereuromark** [comment]
I also bumped PHP to 5.5+ 
Tests are passing now.

**Comment by ADmad** [comment]
In my experience pdf templates are quite simple and personally I actually don't want extra helpers or helper customizations done in `AppView` to be loaded for pdf templates. For e.g. the site could be using bootstrap and `BootstrapUI.HtmlHelper` might be loaded instead of core `HtmlHelper`; I wouldn't want that for pdf templates.

So while making `PdfView` extend `AppView` might be useful to some it can create problem for others, hence I am not in favor of this change.

You always use a custom view class which extends `PdfView` and override it's `initialize()` method. As I mentioned in an earlier comment we can also do something similar to what's done in `BootstrapUI` and put the customizations done in `PdfView` in a trait which would allow greater flexibility on how to customize/compose their custom view class.

**Comment by dereuromark** [comment]
My experience so far 100% always needed and expected the same helpers as the original view, because if you render a view.ctp now as /pdf/view.ctp it sure does need the same helpers to not 500 on the output.
So this is the most common use case when rendering the same HTML content as PDF.

But I understand that there are other use cases, so allowing for them as well sounds reasonable.

**Comment by ADmad** [comment]
One can now use the view builder in controller to configure `PdfView` with required helpers. As explained in previous comment having `PdfView` extends `AppView` could also cause problems.

**Comment by mirko-pagliai** [comment]
I apologize for reopening such an old issue, but I ran into this very problem.

What is currently the best way to do this?

Something like (controller) this?

```php
    public function beforeRender(EventInterface $event)
    {
        if ($this->viewBuilder()->getClassName() === PdfView::class) {
            $this->viewBuilder()->addHelper('Html', ['className' => MyCustomHtmlHelper::class]);
        }

        parent::beforeRender($event);
    }
```

In my case it works fine, although I wondered if it could be done better.

Perhaps it should be documented in the README (...?).

**Comment by ADmad** [comment]
/

**Pull request closed without merging**